File: src/main/java/org/dementhium/content/areas/Area.java
package org.dementhium.content.areas;

import org.dementhium.model.Location;
import org.dementhium.model.player.Player;

/**
 * Represents an area in the world.
 * @author Stephen
 */
public abstract class Area {
    public final String name;
    public final int swX, swY, nwX, nwY, radius, centerX, centerY;
    public transient int length, width;
    private final boolean isPvpZone, isPlusOne, canTele;
    
    public Area(String name, int[] coords, int radius, int centerX, int centerY, boolean isPvpZone, boolean isPlusOne, boolean canTele) {//perhaps something like
        this.name = name;
        this.swX = coords[0];
        this.swY = coords[1];
        this.nwX = coords[2];
        this.nwY = coords[3];
        this.radius = radius;
        this.centerX = centerX;
        this.centerY = centerY;
        this.isPvpZone = isPvpZone;
        this.isPlusOne = isPlusOne;
        this.canTele = canTele;
        this.length = nwX - swX;
        this.width = nwY - swY;
    }
    
    public Area() {
        this.name = "";
        this.nwX = -1;
        this.nwY = -1;
        this.swX = -1;
        this.swY = -1;
        this.radius = -1;
        this.centerX = -1;
        this.centerY = -1;
        this.isPvpZone = false;
        this.isPlusOne = false;
        this.canTele = true;
        this.length = nwX - swX;
        this.width = nwY - swY;
    }
    
    public abstract boolean contains(Location pos);

    public boolean isPvpZone() {
        return isPvpZone;
    }

    public boolean isPlusOneZone() {
        return isPlusOne;
    }

    public boolean canTele() {
        return canTele;
    }

    public String getName() {
        return name;
    }

	public void teleTo(Player player) {
			player.teleport(swX + ((nwX - swX) / 2), swY + ((nwY - swY) / 2), 0);
		
	}

}

File: src/main/java/org/dementhium/content/areas/Area.kt
package org.dementhium.content.areas

import org.dementhium.model.Location
import org.dementhium.model.player.Player

abstract class Area(
   @JvmField val name: String,
   @JvmField val coords: IntArray,
   @JvmField val radius: Int,
   @JvmField val centerX: Int,
   @JvmField val centerY: Int,
   @JvmField val isPvpZone: Boolean,
   @JvmField val isPlusOne: Boolean,
   @JvmField val canTele: Boolean
) {
    @JvmField val swX: Int = coords[0]
    @JvmField val swY: Int = coords[1]
    @JvmField val nwX: Int = coords[2]
    @JvmField val nwY: Int = coords[3]

    @Transient @JvmField
    var length: Int = nwX - swX

    @Transient @JvmField
    var width: Int = nwY - swY

    constructor(): this("",  intArrayOf(-1, -1, -1, -1), -1, -1, -1, false, false, true)

    abstract fun contains(pos: Location): Boolean

    fun canTele(): Boolean  = canTele
    fun getName() = name

    fun teleTo(player: Player) {
        player.teleport(swX + ((nwX - swX) / 2), swY + ((nwY - swY) / 2), 0)
    }

    companion object {
        val InvalidArea = object: Area(
            "INVALID_AREA_XX",
            intArrayOf(-1, -1, -1, -1), 0, 0, 0,
            false, false, false
        ) { override fun contains(pos: Location): Boolean = false }
    }
}